creds/vault/cache.go

```diff
@@ -0,0 +1,116 @@
+package vault
+
+import (
+	"sync"
+	"time"
+
+	vaultapi "github.com/hashicorp/vault/api"
+)
+
+type cachedSecret struct {
+	deadline time.Time
+	secret   *vaultapi.Secret
+}
+
+// A Cache caches secrets read from a SecretReader until the lease on
+// the secret expires. Once expired the credential is proactively
+// deleted from cache to maintain a smaller cache footprint.
+type Cache struct {
+	sync.RWMutex
+	cache    map[string]*cachedSecret
+	newItems chan time.Time
+	sr       SecretReader
+}
+
+// TODO: Should a cache have a max size to
+// prevent unbounded growth?
```

Comment:
Hmm, yeah this could be really interesting for large multi-tenant Concourse installations. Some credentials can be fairly large (e.g. private keys), so it could add up quickly. You could always just not use caching, but chances are you probably *want* caching in that same scenario, so it becomes a question of which you sacrifice: ATC RAM vs. Vault CPU.

I think it's fine to ship without this, but maybe the thing to do here is just have a configurable max retention for each cache? This could either be time-based or usage based (i.e. each time a credential's cache is hit, we keep it for a bit longer, still limited by its expiry obviously).